File: projeto/frontend/site/vox/src/app/services/event.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface EventCategory { id: number; name: string; description?: string; }
export interface EventImage { id: number; eventId?: number; url: string; }
export interface CivicEvent {
  id: number;
  title: string;
  description?: string;
  categoryId: number;
  price?: number | null;
  startDate?: string | null;
  endDate?: string | null;
  neighborhood?: string | null;
  street?: string | null;
  number?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  /** Legacy payload compatibility. */
  location?: string | null;
  municipalityId?: number;
  authorId?: number;
  createdAt?: string;
  images?: EventImage[];
}
export interface EventPage { content: CivicEvent[]; page: number; size: number; totalElements: number; }
export interface EventFilters {
  search?: string; categoryId?: number; municipalityId?: number; free?: boolean; hasImage?: boolean;
  startFrom?: string; startTo?: string; page?: number; size?: number;
}

@Injectable({ providedIn: 'root' })
export class EventService {
  private readonly apiUrl = 'http://localhost:8080/api';
  constructor(private http: HttpClient) {}

  getCategories(): Observable<EventCategory[]> {
    return this.http.get<EventCategory[]>(`${this.apiUrl}/event-categories`);
  }

  getEvents(filters: EventFilters = {}): Observable<EventPage> {
    let params = new HttpParams();
    Object.entries({ size: 12, page: 0, ...filters }).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') params = params.set(key, String(value));
    });
    return this.http.get<EventPage>(`${this.apiUrl}/events`, { params });
  }

  getEvent(id: number): Observable<CivicEvent> {
    return this.http.get<CivicEvent>(`${this.apiUrl}/events/${id}`);
  }

  getEventImages(id: number): Observable<EventImage[]> {
    return this.http.get<EventImage[]>(`${this.apiUrl}/events/${id}/images`);
  }

  createEvent(data: Record<string, string>, file?: File): Observable<void> {
    return this.http.post(`${this.apiUrl}/events`, this.formData(data, file), { responseType: 'text' }) as unknown as Observable<void>;
  }

  updateEvent(id: number, data: Record<string, string>, file?: File): Observable<void> {
    return this.http.put(`${this.apiUrl}/events/${id}`, this.formData(data, file), { responseType: 'text' }) as unknown as Observable<void>;
  }

  deleteEvent(id: number): Observable<void> {
    return this.http.delete(`${this.apiUrl}/events/${id}`, { responseType: 'text' }) as unknown as Observable<void>;
  }

  private formData(data: Record<string, string>, file?: File): FormData {
    const form = new FormData();
    Object.entries(data).forEach(([key, value]) => { if (value !== '') form.append(key, value); });
    if (file) form.append('file', file, file.name);
    return form;
  }
}

File: projeto/frontend/app/app_vox/lib/services/event_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/event.dart';
import 'api_client.dart';

class EventService {
  static const _base = '${ApiClient.baseUrl}/api';
  Future<EventPage> getEvents({
    String? search,
    int? categoryId,
    int? municipalityId,
    bool free = false,
    int page = 0,
    int size = 12,
  }) async {
    final query = <String, String>{'page': '$page', 'size': '$size'};
    if (search != null && search.trim().isNotEmpty) {
      query['search'] = search.trim();
    }
    if (categoryId != null) query['categoryId'] = '$categoryId';
    if (municipalityId != null) query['municipalityId'] = '$municipalityId';
    if (free) query['free'] = 'true';
    final response = await http.get(
      Uri.parse('$_base/events').replace(queryParameters: query),
      headers: await ApiClient.authHeaders(json: false),
    );
    ApiClient.checkResponse(response);
    return EventPage.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<CivicEvent> getEvent(int id) async {
    final response = await http.get(
      Uri.parse('$_base/events/$id'),
      headers: await ApiClient.authHeaders(json: false),
    );
    ApiClient.checkResponse(response);
    return CivicEvent.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<List<EventImage>> getEventImages(int id) async {
    final response = await http.get(
      Uri.parse('$_base/events/$id/images'),
      headers: await ApiClient.authHeaders(json: false),
    );
    ApiClient.checkResponse(response);
    return (jsonDecode(response.body) as List<dynamic>)
        .whereType<Map<String, dynamic>>()
        .map(EventImage.fromJson)
        .toList();
  }

  Future<List<Map<String, dynamic>>> getCategories() async {
    final response = await http.get(
      Uri.parse('$_base/event-categories'),
      headers: await ApiClient.authHeaders(json: false),
    );
    ApiClient.checkResponse(response);
    return (jsonDecode(response.body) as List<dynamic>)
        .whereType<Map<String, dynamic>>()
        .toList();
  }

  Future<int?> createEvent(Map<String, String> fields) async {
    final streamed = await ApiClient.multipartRequest(
      'POST',
      '$_base/events',
      _nonEmptyFields(fields),
    );
    final response = await http.Response.fromStream(streamed);
    ApiClient.checkResponse(response);
    return ApiClient.locationId(response);
  }

  Future<void> updateEvent(int id, Map<String, String> fields) async {
    final streamed = await ApiClient.multipartRequest(
      'PUT',
      '$_base/events/$id',
      _nonEmptyFields(fields),
    );
    ApiClient.checkResponse(await http.Response.fromStream(streamed));
  }

  Future<void> deleteEvent(int id) async {
    final response = await http.delete(
      Uri.parse('$_base/events/$id'),
      headers: await ApiClient.authHeaders(json: false),
    );
    ApiClient.checkResponse(response);
  }

  Map<String, String> _nonEmptyFields(Map<String, String> fields) =>
      Map.fromEntries(
        fields.entries.where((entry) => entry.value.trim().isNotEmpty),
      );
}
